app/src/main/java/io/appium/uiautomator2/handler/GetElementAttribute.java

```diff
@@ -79,70 +90,142 @@ public AppiumResponse safeHandle(IHttpRequest request) {
+            throw new UiAutomator2Exception("Could not retrieve accessibility scroll data; unable to determine scrollable offset");
         }
+
+        // if we're in some views, like ScrollViews, we get x/y values directly, and we can simply return
+        if (lastScrollData.getMaxScrollY() != -1) {
+            return lastScrollData.getMaxScrollY();
+        }
+
+        // in other views, like List or Grid, we get item counts and indexes, and we have to turn
+        // that into pixels by doing some math
+        if (lastScrollData.getItemCount() == -1) {
+            throw new UiAutomator2Exception("Did not get either scrollY or itemCount from accessibility scroll data");
+        }
+
+        return getScrollableOffsetByItemCount(uiScrollable, lastScrollData.getItemCount());
     }
 
-    private static int getScrollableOffset(AndroidElement uiScrollable) throws UiObjectNotFoundException, ClassNotFoundException, InvalidSelectorException {
-        AccessibilityNodeInfo nodeInfo = null;
-        int offset = 0;
-        Object actualObject = uiScrollable.getUiObject();
-        if (actualObject instanceof UiObject) {
-            UiObject object = (UiObject) uiScrollable.getChild(new UiSelector().index(0));
-            Method findAccessibilityNodeInfoMethod = ReflectionUtils.method(UiObject.class, "findAccessibilityNodeInfo", long.class);
-            long waitForSelectorTimeout = (long) ReflectionUtils.getField(UiObject.class, "WAIT_FOR_SELECTOR_TIMEOUT", object);
-
-            nodeInfo = (AccessibilityNodeInfo) ReflectionUtils.invoke(findAccessibilityNodeInfoMethod, object, waitForSelectorTimeout);
-        } else {
-            UiObject2 childObject = ((UiObject2) actualObject).getChildren().get(0);
-            try {
-                nodeInfo = AccessibilityNodeInfoGetter.fromUiObject(childObject);
-            } catch (UiAutomator2Exception ignored) {
+    private static int getScrollableOffsetByItemCount (AndroidElement uiScrollable, int itemCount) {
+        Logger.debug("Figuring out scrollableOffset via item count");
+        Object scrollObject = uiScrollable.getUiObject();
+
+        // here we loop through the children and get their bounds until the height differs, then
+        // regardless of whether we have a list or a grid, we'll know the height of an item/row
+        try {
+            int lastExaminedYInitVal = -999999; // something that could never be a 'top'
+            int itemsPerRow = 0;
+            int rowHeight = 0;
+            int lastExaminedItemY = lastExaminedYInitVal;
+            int numRowsExamined = 0;
+            int numRowsToExamine = 3; // examine a few rows since the top ones often have bad offsets
+            Object lastExaminedItem = null;
+
+            UiObjectChildGenerator gen = new UiObjectChildGenerator(scrollObject);
+            for (Object item : gen) {
+                Rect bounds = getElementBoundsInScreen(item);
+
+                if (item == null) {
```

Comment:
should this check be placed before getElementBoundsInScreen ?

Reply:
yes, yes it should :-)